File: tools/ExternalEvaluation/Java/ExternalEvaluation.Service/src/com/heuristiclab/problems/externalevaluation/PollService.java
package com.heuristiclab.problems.externalevaluation;

import java.io.*;
import java.util.*;
import java.util.concurrent.*;
import com.heuristiclab.problems.externalevaluation.ExternalEvaluationMessages.*;
import com.heuristiclab.problems.externalevaluation.ExternalEvaluationMessages.QualityMessage.Type;

public class PollService extends Service {
	private LinkedBlockingQueue<SolutionMessage> messageQueue;
	private ArrayList<SolutionEventListener> listenerList;
	
	public PollService(IListenerFactory listenerFactory, int nThreads) {
		super(listenerFactory, nThreads);
		messageQueue = new LinkedBlockingQueue<SolutionMessage>();
		listenerList = new ArrayList<SolutionEventListener>();
	}
	
	@Override
	public void stop() {
		messageQueue.clear();
		super.stop();
	}
	
	public synchronized SolutionMessage getSolution() {
		try {
			return messageQueue.take();
		} catch (InterruptedException e) {
			e.printStackTrace();
		}
		return null;
	}
	
	public void sendQualityMessage(SolutionMessage sMsg, QualityMessage qMsg) throws IOException {
		send(qMsg, sMsg);
	}
	
	public QualityMessage.Builder prepareQualityMessage(SolutionMessage msg, double quality) {
		return QualityMessage.newBuilder()
			.setSolutionId(msg.getSolutionId())
			.setType(Type.SingleObjectiveQualityMessage)
			.setExtension(
					SingleObjectiveQualityMessage.qualityMessage, 
					SingleObjectiveQualityMessage.newBuilder().setQuality(quality).build());
	}
		
	public void sendQuality(SolutionMessage msg, double quality) throws IOException {
		send(prepareQualityMessage(msg, quality).build(), msg);
	}
	
	public QualityMessage.Builder prepareQualitiesMessage(SolutionMessage msg, double[] qualities) {
		List<Double> qualitiesList = new ArrayList<Double>(qualities.length);
		for (double quality : qualities) qualitiesList.add(quality);
		return QualityMessage.newBuilder()
				.setSolutionId(msg.getSolutionId())
				.setType(Type.MultiObjectiveQualityMessage)
				.setExtension(
						MultiObjectiveQualityMessage.qualityMessage, 
						MultiObjectiveQualityMessage.newBuilder().addAllQualities(qualitiesList).build());
	}
		
	public void sendQualities(SolutionMessage msg, double[] qualities) throws IOException {
		send(prepareQualitiesMessage(msg, qualities).build(), msg);
	}
	
	public void addSolutionEventListener(SolutionEventListener l) {
		synchronized(listenerList) {
			listenerList.add(l);
		}
	}
	public void removeSolutionEventListener(SolutionEventListener l) {
		synchronized(listenerList) {
			listenerList.remove(l);
		}
	}
	
	@Override
	protected void onSolutionProduced(SolutionMessage msg, IChannel channel) {
		super.onSolutionProduced(msg, channel);
		try {
			messageQueue.put(msg);
		} catch (InterruptedException e) {
			e.printStackTrace();
			return;
		}
		synchronized(listenerList) {
			Iterator<SolutionEventListener> listeners = listenerList.iterator();
			while( listeners.hasNext() ) {
				listeners.next().solutionReceived(new EventObject(this));
			}
		}
	}
}

File: tools/ExternalEvaluation/CSharp/ExternalEvaluation.Service/PollService.cs
#region License Information
/* HeuristicLab
 * Copyright (C) 2002-2010 Heuristic and Evolutionary Algorithms Laboratory (HEAL)
 *
 * This file is part of HeuristicLab.
 *
 * HeuristicLab is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * HeuristicLab is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with HeuristicLab. If not, see <http://www.gnu.org/licenses/>.
 */
#endregion

using System;

namespace HeuristicLab.Problems.ExternalEvaluation.Service {
  public class PollService : Service {
    private BlockingQueue<SolutionMessage> messageQueue;

    public PollService(IListenerFactory driverFactory)
      : base(driverFactory) {
      messageQueue = new BlockingQueue<SolutionMessage>(int.MaxValue);
    }

    public override void Stop() {
      messageQueue.Close();
      base.Stop();
    }

    public SolutionMessage GetSolution() {
      try {
        return messageQueue.Dequeue();
      } catch (QueueClosedException) { }
      return null;
    }

    public QualityMessage.Builder PrepareQualityMessage(SolutionMessage msg, double quality) {
      QualityMessage.Builder qualityMessageBuilder = QualityMessage.CreateBuilder();
      qualityMessageBuilder.SetSolutionId(msg.SolutionId)
                           .SetQuality(quality);
      return qualityMessageBuilder;
    }

    public void SendQualityMessage(SolutionMessage sMsg, QualityMessage qMsg) {
      Send(qMsg, sMsg);
    }

    public void SendQuality(SolutionMessage msg, double quality) {
      QualityMessage.Builder qualityMessageBuilder = QualityMessage.CreateBuilder();
      qualityMessageBuilder.SetSolutionId(msg.SolutionId)
                           .SetQuality(quality);
      Send(qualityMessageBuilder.Build(), msg);
    }

    public event EventHandler SolutionProduced;

    protected override void OnSolutionProduced(SolutionMessage msg, IChannel channel) {
      base.OnSolutionProduced(msg, channel);
      try {
        messageQueue.Enqueue(msg);
      } catch (QueueClosedException) {
        return;
      }
      EventHandler handler = SolutionProduced;
      if (handler != null) handler(this, EventArgs.Empty);
    }
  }
}
